frame/staking/src/lib.rs

```diff
@@ -819,8 +896,49 @@ decl_module! {
 
 		fn deposit_event() = default;
 
-		fn on_initialize() {
-			Self::ensure_storage_upgraded();
+		fn on_initialize(now: T::BlockNumber) {
+			if
+				// if we don't have any ongoing offchain compute.
+				Self::election_status() == OffchainElectionStatus::None &&
```

Comment:
Why is this logic running on-chain? It should just happen within `offchain` worker, and `election_status` should only have  two values (`None` and `Received`). Or possibly this is also not necessary, since we can just check the `QueuedElected`.

Reply:
The main reason is that currently I assumed that the chain dictates _when one is allowed to submit_, hence the chain must know this. It can either compute it per block, or, as I have done now, store one state variable and do less computation. I am essentially favouring Memory over CPU here. 

Other than that, `OffchainElectionStatus` is redundant by nature, because it can always be safely computed by everyone.  The combination of checking if we are in the correct time span `AND` the value of   `QueuedElected ` as you mentioned is enough. The other reasons I let it exist for now is to just be more explicit (since I suppose it is not a super heavy structure to store on-chain). 

Removing it should naturally occur as the implementation matures.

Reply:
In the other comment below, I actually explained why this should actually stay on-chain.